### PR TITLE
Syntax highlighting (#4 & #11)

### PR DESCRIPTION
**Pull request by douweschulte**
This is my current state of the implementation of #4 and #11.
image
Merge needs to wait until these PRs are merged in Enselic/public_items:
* Enselic/public_items#53 with the support for tokenization.
* Enselic/public_items#63 with the support for token based diffing.

### CODE REVIEW

File: src/markdown.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ pub fn print_diff(w: &mut impl std::io::Write, diff: &PublicItemsDiff) -> std::i
         w,
         "## Added items to the public API",
         &diff.added,
-        |w, item| writeln!(w, "* `{}`", item),
+        |w, item| writeln!(w, "* {}", print_with_colour(item)),
```
Comment (Enselic) [inline src/markdown.rs]: Maybe we should output regular (colored) text by default and make markdown opt-in? I'm not sure it makes sense to put colors in markdown. Let me know if you want me to write the code to select output format so you can focus on tokenization and coloring.

Comment (douweschulte) [reply src/markdown.rs]: That sounds lovely. If you have an idea how to tackle this, that would be very nice. Otherwise I will work on that after #11.

Comment (Enselic) [reply src/markdown.rs]: Here is the PR (I pinged you in it): https://github.com/Enselic/cargo-public-items/pull/14

**Comment by douweschulte** [comment]
Sneek peak:
syntax_highlighting

I am quite happy with this. I need to test how it handles other inputs so that needs some time. After that I will go and clean up the stuff that is not needed. I envision to remove the prefix/path/suffix from the `PublicItem` if possible en replace it with the fully tokenized output, let me know what you think about that.

**Comment by Enselic** [comment]
Nice! Colors makes such a big difference. Looking forward to the next update.

> I envision to remove the prefix/path/suffix from the PublicItem if possible en replace it with the fully tokenized output, let me know what you think about that.

That sounds perfectly reasonable on a conceptual level, so please go ahead with that 👍 I'm happy to answer any questions that might arise from such an operation.

**Comment by douweschulte** [comment]
image

I made a start on #11 see above. The code is not really clean yet. I am not really sure how to style it in a way that works best for other people, so if you have ideas that would be welcome.

**Comment by Enselic** [comment]
Yeah it's not obvious how to present diffs in a good way. Maybe it would make sense to get regular coloring in place first, and then after that we can iterate on how to show diffs?

In your example above, I think one improvement would be to have the green color on the + line, and red color on the - line. So the added `impl` should have green background, and the removed `String,` should have red background, rather than having the color on the "opposite" side.

Another source to draw inspiration from could be `git show/diff --color-words`, but I'm not sure I like that style:

Skärmavbild 2022-04-02 kl  09 36 58

**Comment by douweschulte** [comment]
image
Okay this looks way better, the background colour does not work that well because it is way to bright and I do not see a way to get dimmed background colours. So I am thinking about doing other style things, like strikethrough or italic. But this means that the `tokens::ChangedTokenStream` is pretty much where I want it.

**Comment by Enselic** [comment]
Hi again! Just an update. I have now
* Renamed public_items to public-api
* Released public-api v0.9.0 that contains your new tokens API
* Released cargo-public-api v0.9.0 that depends on the new public-api lib, but does not make use of tokens yet.

It would be fantastic if you could create a new PR (I think that will be more clean than updating this one) that gets syntax highlighting in place. I would really like us to do that separately from the intra-item diffing, so keep things focused and incremental. Again, no rush, feel free to take your time with that :)

**Comment by douweschulte** [comment]
Okay sounds like a plan. I will get on it when I have time.

**Pull request closed without merging**